### ISSUE TITLE
Query: incorrect data returned for queries projecting a single constant from a subquery

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#11459]
Repro using GearsOfWar model:

```cs
ctx.Squads.Select(s => new { s.Name, Gear = s.Members.Where(g => g.HasSoulPatch).Select(g => 42).FirstOrDefault() })
```

query plan:
```
(QueryContext queryContext) => IEnumerable<<>f__AnonymousType283<string, int>> _InterceptExceptions(
    source: IEnumerable<<>f__AnonymousType283<string, int>> _ShapedQuery(
        queryContext: queryContext, 
        shaperCommandContext: SelectExpression: 
            SELECT [s].[Name], (
                SELECT TOP(1) 1
                FROM [Gears] AS [g]
                WHERE ([g].[Discriminator] IN (N'Officer', N'Gear') AND ([g].[HasSoulPatch] = 1)) AND ([s].[Id] = [g].[SquadId])
            ) AS [Gear]
            FROM [Squads] AS [s], 
        shaper: TypedProjectionShaper<ValueBufferShaper, ValueBuffer, <>f__AnonymousType283<string, int>>), 
    contextType: TestModels.GearsOfWarModel.GearsOfWarContext, 
    logger: DiagnosticsLogger<Query>, 
    queryContext: queryContext)
```

42 is supposed to be client side projection but its being "swallowed" and we end up projecting 1 instead

If the constant if of a different type than int we get a runtime error in type mapping:

```
   at Microsoft.EntityFrameworkCore.Storage.RelationalTypeMappingSourceExtensions.GetMapping(IRelationalTypeMappingSource typeMappingSource, Type clrType)
   at Microsoft.EntityFrameworkCore.Storage.TypeMaterializationInfo..ctor(Type modelClrType, IProperty property, IRelationalTypeMappingSource typeMappingSource, Int32 index)
   at Microsoft.EntityFrameworkCore.Query.Expressions.SelectExpression.<GetMappedProjectionTypes>b__70_0(Expression e)
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Expressions.SelectExpression.<GetMappedProjectionTypes>d__70.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.CreateValueBufferFactory(IRelationalValueBufferFactoryFactory relationalValueBufferFactoryFactory, DbDataReader dataReader)
   at Microsoft.EntityFrameworkCore.Query.Internal.ShaperCommandContext.<NotifyReaderCreated>b__14_0(FactoryAndReader s)
   at Microsoft.EntityFrameworkCore.Internal.NonCapturingLazyInitializer.EnsureInitialized[TParam,TValue](TValue& target, TParam param, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Query.Internal.ShaperCommandContext.NotifyReaderCreated(DbDataReader dataReader)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at 
```

We should fully translate projections for subqueries

**Comment by maumar** [comment]
related to https://github.com/aspnet/EntityFrameworkCore/issues/10946

**Comment by smitpatel** [comment]
If we translate this also enable aggregate over a constant in group by case.

**Comment by smitpatel** [comment]
@maumar - Can you verify this and close if fixed? (Also make sure that group by now translates).

**Comment by maumar** [comment]
fixed in 3902d943a489107d9e2336a434ecf489376d0516
